### ISSUE TITLE
[Bug] Topic Operator Reconcilation Revert Kafkatopic CRD Definitions

### ISSUE THREAD

**Issue by kenju** [strimzi/strimzi-kafka-operator#4998]
_Please use this to only for bug reports. For questions or when you need help, you can use the [GitHub Discussions](https://github.com/strimzi/strimzi-kafka-operator/discussions), our [#strimzi Slack channel](https://slack.cncf.io/) or out [user mailing list](https://lists.cncf.io/g/cncf-strimzi-users/topics)._

**Describe the bug**

Here is the `topic.yaml` we'd like to appy:

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: KafkaTopic
metadata:
  labels:
    strimzi.io/cluster: kafka-cluster
  name: foo
spec:
  config:
    cleanup.policy: delete
    min.insync.replicas: "2"
    retention.ms: "2592000000"
  partitions: 12
  replicas: 3
  topicName: foo
```

Let's say, there is a diff in `min.insync.replicas`:

```sh
kubectl diff -f ./topic.yaml
```

```diff
-  generation: 78
+  generation: 79
@@ -16,6 +16,7 @@
 spec:
   config:
     cleanup.policy: delete
+    min.insync.replicas: "2"
     retention.ms: "2592000000"
   partitions: 12
   replicas: 3
```

Let's apply the CRD. After applying this file, it shows no diff.

```sh
$ kubectl apply -f ./topic.yaml
kafkatopic.kafka.strimzi.io/foo configured
$ kubectl diff -f ./topic.yaml
(nothing)
```

Let's wait for the `reconciliationIntervalSeconds` time, and topic-operator will try to synchronize `Kafkatopic` CRDs and the actual kafka topics... 

```
{"message":"2021-05-19 14:07:39,34541 INFO  [vert.x-eventloop-thread-1] TopicOperator:1417 - Success reconciling KafkaTopic foo/foo"}
```

**After that, we can confirm that `kubectl diff` shows the diff again. Notice that `generation` is also incremented to `81`.**

```sh
$ kubectl diff -f ./topic.yaml
```

```diff
-  generation: 80
+  generation: 81
@@ -16,6 +16,7 @@
 spec:
   config:
     cleanup.policy: delete
+    min.insync.replicas: "2"
     retention.ms: "2592000000"
   partitions: 12
   replicas: 3
```

When we check the actual configuration using [Kafka Tools](https://github.com/confluentinc/kafka/tree/master/bin), topic-operator seems to revert our changes:

```sh
./bin/kafka-configs \
  --command-config ./client.properties \
    --bootstrap-server foo.example.com:9999 \
    --entity-type topics \
    --entity-name foo \
    --describe --all
```

```
---%<---
  min.insync.replicas=1 sensitive=false synonyms={DEFAULT_CONFIG:min.insync.replicas=1}
---%<---
```

As far as I understand, topic-operator stores topic state in it's internal topics. I wonder if the state in those topics are broken somehow, then I tried to look into the topics but could not find out via `bin/kafka-topics.sh --list`:

https://strimzi.io/docs/operators/latest/using.html#ref-topic-operator-store-str

> __strimzi_store_topic
> __strimzi-topic-operator-kstreams-topic-store-changelog

**To Reproduce**
1. Deploy topic-operator
2. Create Custom Resource 'Kafkatopic' and apply
3. Wait for a while until the next reconcilatoin happens
4. See the configuration in `Kafkatopic` CRDs are reverted

**Expected behavior**

**Environment (please complete the following information):**
 - Strimzi version: 0.21.1
 - Installation method: YAML files
 - Kubernetes cluster: Kubernetes
 - Infrastructure: Amazon EKS

Also, topic-operator is deployed via `Kafka` spec, not as a standalone.

**YAML files and logs**

Attached in the above.

**Additional context**

It happens on multiple clusters, with completely the same setup (YAML Files, version, infrastructure).

Based on [how topic-operator works](https://strimzi.io/docs/operators/latest/using.html#ref-topic-operator-store-str), my mental model says something is going wrong in the topic store and its state, but I do not know how to dig deeper.

Do you have any ideas why it's happening? Thank you.

**Comment by scholzj** [comment]
I'm not sure that would be the issue. Are you sure you don't have some application changing it back directly in Kafka? In any case, if you can provide log from the Topic Operator (ideally on debug level) it should show what is happening.

**Comment by kenju** [comment]
@scholzj Thanks for looking into it.

>  Are you sure you don't have some application changing it back directly in Kafka? 

We were running MirrorMaker 2.0, but we are sure that the topic configuration on the source cluster and our Kafkatopic CRDs have the completely same configuration. Also, the issue has been still happening even after we stopped running MM2.

> In any case, if you can provide log from the Topic Operator (ideally on debug level) it should show what is happening.

Let me get back to you shortly. I'll dry debug log level :+1:

**Comment by scholzj** [comment]
MM2 normally doesn't mirror some parts of topic configuration such as replication factor etc. So in theory it could be an issue ... but if the same was happening after you stopped it then I guess it was not it. The log might tell us more.

**Comment by kenju** [comment]
@scholzj 

> Also, the issue has been still happening even after we stopped running MM2.

> MM2 normally doesn't mirror some parts of topic configuration such as replication factor etc. So in theory it could be an issue 

You were right. In more precisely, we did not stopped "running" MM2 but we just stopped "mirroring" source topics (by making changes to MM2 configuration). After stopping the MM2 process, we find that no more diff appeared.

Considering above, I can confirm that MM2 had been updating the actual topic configuration simultaneously. Let me close this issue now, and thanks for your help.
